agent/elia_chat/widgets/chatbox.py

```diff
@@ -0,0 +1,384 @@
+            cursor_location,
+            half_height,
+        )
+        self.scroll_relative(y=half_height, animate=False)
+        self.move_cursor(target)
+
+    def action_cursor_half_page_up(self) -> None:
+        """Move the cursor and scroll down half of a page."""
+        half_height = self.content_size.height // 2
+        _, cursor_location = self.selection
+        target = self.navigator.get_location_at_y_offset(
+            cursor_location,
+            -half_height,
+        )
+        self.scroll_relative(y=-half_height, animate=False)
+        self.move_cursor(target)
+
+
+class Chatbox(Widget, can_focus=True):
+    BINDINGS: ClassVar[list[BindingType]] = [
+        Binding(key="up,k", action="up", description="Up", show=False),
+        Binding(key="down,j", action="down", description="Down", show=False),
+        Binding(key="enter", action="select", description="Toggle select mode"),
+        Binding(
+            key="y,c",
+            action="copy_to_clipboard",
+            description="Copy full message",
+            key_display="y",
+        ),
+        Binding(
+            key="escape",
+            action="screen.focus('prompt')",
+            description="Focus prompt",
+            key_display="esc",
+        ),
+    ]
+
+    class CursorEscapingBottom(Message):
+        """Sent when the cursor moves down from the bottom message."""
+
+    selection_mode = reactive(False, init=False)
+
+    def __init__(
+        self,
+        message: ChatMessage,
+        model: EliaChatModel,
+        name: str | None = None,
+        id: str | None = None,
+        classes: str | None = None,
+        disabled: bool = False,
+    ) -> None:
+        super().__init__(
+            name=name,
+            id=id,
+            classes=classes,
+            disabled=disabled,
+        )
+        self.message = message
+        self.model = model
+
+    def on_mount(self) -> None:
+        litellm_message = self.message.message
+        role = litellm_message["role"]
+        if role == "assistant":
+            self.add_class("assistant-message")
+            self.border_title = "Agent"
+        else:
+            self.add_class("human-message")
+            self.border_title = "You"
+
+    def action_up(self) -> None:
+        self.screen.focus_previous(Chatbox)
+
+    def action_down(self) -> None:
+        if self.parent and self is self.parent.children[-1]:
+            self.post_message(self.CursorEscapingBottom())
+        else:
+            self.screen.focus_next(Chatbox)
+
+    def action_select(self) -> None:
+        self.selection_mode = not self.selection_mode
+        self.set_class(self.selection_mode, "selecting")
+
+    def action_copy_to_clipboard(self) -> None:
+        if not self.selection_mode:
+            text_to_copy = self.message.message.get("content")
+            if isinstance(text_to_copy, str):
+                try:
+                    pyperclip.copy(text_to_copy)
```

Comment:
Use pyperclip when inside of a mac terminal. Otherwise use `self.app.copy_to_clipboard(text_to_copy)` when in ssh terminals.

See: https://darren.codes/posts/textual-copy-paste/